src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Integer/DenseIntegralFrozenDictionary.cs

```diff
@@ -0,0 +1,234 @@
+                return TryCreate<TKey, int, TValue>(source, out result);
+
+            result = null;
+            return false;
+        }
+
+        private static bool TryCreate<TKey, TKeyUnderlying, TValue>(Dictionary<TKey, TValue> source, [NotNullWhen(true)] out FrozenDictionary<TKey, TValue>? result)
+            where TKey : notnull
+            where TKeyUnderlying : unmanaged, IBinaryInteger<TKeyUnderlying>
+        {
+            // Start enumerating the dictionary to ensure it has at least one element.
+            Dictionary<TKey, TValue>.Enumerator e = source.GetEnumerator();
+            if (e.MoveNext())
+            {
+                // Get that element and treat it as the min and max. Then continue enumerating the remainder
+                // of the dictionary to count the number of elements and track the full min and max.
+                int count = 1;
+                int min = int.CreateTruncating((TKeyUnderlying)(object)e.Current.Key);
+                int max = min;
+                while (e.MoveNext())
+                {
+                    count++;
+                    int key = int.CreateTruncating((TKeyUnderlying)(object)e.Current.Key);
+                    if (key < min)
+                    {
+                        min = key;
+                    }
+                    else if (key > max)
+                    {
+                        max = key;
+                    }
+                }
+
+                // Based on the min and max, determine the spread. If the range fits within a non-negative Int32
+                // and the ratio of the number of elements in the dictionary to the length is within the allowed
+                // threshold, create the new dictionary.
+                long length = (long)max - min + 1;
+                Debug.Assert(length > 0);
+                if (length <= int.MaxValue &&
+                    (double)count / length >= CountToLengthRatio)
+                {
+                    // Create arrays of the keys and values, sorted ascending by key.
+                    var keys = new TKey[count];
+                    var values = new TValue[keys.Length];
+                    int i = 0;
+                    foreach (KeyValuePair<TKey, TValue> entry in source)
+                    {
+                        keys[i] = entry.Key;
+                        values[i] = entry.Value;
+                        i++;
+                    }
+
+                    if (i != keys.Length)
+                    {
+                        throw new InvalidOperationException(SR.CollectionModifiedDuringEnumeration);
+                    }
+
+                    // Sort the values so that we can more easily check for contiguity but also so that
+                    // the keys/values returned from various properties/enumeration are in a predictable order.
+                    Array.Sort(keys, values);
```

Comment:
Sorting is expensive. Is it necessary?
1. `min == 0 && length == count` is enough to get the result of `isFull`.
2.  It seems no need to sort before creating `WithOptionalValues`. And similar implementation can be used when creating `WithFullValues` to skip sorting.

Reply:
>  Is it necessary?

As noted in the comment, I wanted the keys/values in order, as we do for e.g. SmallValueTypeComparableFrozenDictionary.

Reply:
> And even if we need to order the keys of DenseIntegralFrozenDictionary , there's better way to implement that without sorting.

?

Reply:
For ``
```csharp
int keyIndex = 0;
for (i = 0; i < optionalValues.Length; i++)
{
    if (optionalValues[i]. HasValue);
    keys[keyIndex] = i + min;
    values[keyIndex] = optionalValues[I].Value;
}
```

Reply:
But that's not the point. Users never expect `FrozenDictionary.Keys` are sorted.